### PR TITLE
Work-around for net_kernel:get_status() race condition

### PR DESCRIPTION
**Pull request by slfritchie**
The node monitor race in `net_kernel:get_status()` is present
both in all (?) R14B releases as well as R15B.  It's not going to
be fixed anytime quickly.

So, the work-around involves even more knowledge of the internals
of `net_kernel`, but we were already breaking that barrier: in for
a penny, in for a dollar.

Passes basho_expect's regression_az1018 test.

### CODE REVIEW

**Comment by jonmeredith** [comment]
Seen again on a customer system 

```
(riakuser@host-2174.customer.org)12> io:format("~s\n", [element(2, erlang:process_info(pid(0,69,0), backtrace))]).    
Program counter: 0x000000000578ab28 (net_kernel:get_status/3 + 216)
CP: 0x0000000000000000 (invalid)
arity = 0

0xfffffd7fcd7ec500 Return addr 0x000000000578a870 (net_kernel:get_node_info/1 + 424)
y(0)     'riakuser@host-2167.customer.org'
y(1)     <0.120.0>

0xfffffd7fcd7ec518 Return addr 0x000000000578afb8 (net_kernel:get_nodes_info/2 + 96)
y(0)     normal
y(1)     {net_address,{{10,10,25,90},53096},"host-2167.customer.org",tcp,inet}
y(2)     <0.120.0>
y(3)     up

0xfffffd7fcd7ec540 Return addr 0x00000000058e6d38 (riak_sysmon_filter:get_node_map/0 + 64)
y(0)     [{'riakuser@host-2176.customer.org',[{owner,<0.2135.1747>},{state,pending},{address,{net_address,undefined,undefined,tcp,inet}},{type,normal},{in,0},{out,0}]},{'riakuser@host-2173.customer.org',[{owner,<0.80.0>},{state,up},{address,{net_address,{{10,10,25,88},39198},"host-2173.customer.org",tcp,inet}},{type,normal},{in,14824238},{out,17305647}]},{'riakuser@host-2172.customer.org',[{owner,<0.125.0>},{state,up},{address,{net_address,{{10,10,25,86},54204},"host-2172.customer.org",tcp,inet}},{type,normal},{in,11840751},{out,14520696}]},{'riakuser@host-2169.customer.org',[{owner,<0.28785.1702>},{state,up_pending},{address,{net_address,{{10,10,25,112},52001},"host-2169.customer.org",tcp,inet}},{type,normal},{in,0},{out,0}]},{'riakuser@host-2168.customer.org',[{owner,<0.121.0>},{state,up},{address,{net_address,{{10,10,25,105},34719},"host-2168.customer.org",tcp,inet}},{type,normal},{in,10051796},{out,10122352}]},{'riakuser@host-2171.customer.org',[{owner,<0.124.0>},{state,up},{address,{net_address,{{10,10,25,117},44123},"host-2171.customer.org",tcp,inet}},{type,normal},{in,11630726},{out,11447109}]},{'riakuser@host-2170.customer.org',[{owner,<0.123.0>},{state,up},{address,{net_address,{{10,10,25,113},36701},"host-2170.customer.org",tcp,inet}},{type,normal},{in,7679927},{out,7938672}]}]
y(1)     ['riakuser@host-2177.customer.org']
y(2)     'riakuser@host-2167.customer.org'

0xfffffd7fcd7ec560 Return addr 0x00000000058e63c8 (riak_sysmon_filter:handle_info/2 + 2600)
y(0)     Catch 0x00000000058e6db0 (riak_sysmon_filter:get_node_map/0 + 184)

0xfffffd7fcd7ec570 Return addr 0x0000000005690648 (gen_server:handle_msg/5 + 1680)
y(0)     []
y(1)     {0,nil}
y(2)     {state,0,30,0,2,{0,nil},[{{{10,10,25,140},55002},'riakuser@host-2177.customer.org'},{{{10,10,25,90},53096},'riakuser@host-2167.customer.org'},{{{10,10,25,109},55747},'riakuser@host-2176.customer.org'},{{{10,10,25,88},39198},'riakuser@host-2173.customer.org'},{{{10,10,25,86},54204},'riakuser@host-2172.customer.org'},{{{10,10,25,112},52001},'riakuser@host-2169.customer.org'},{{{10,10,25,105},34719},'riakuser@host-2168.customer.org'},{{{10,10,25,117},44123},'riakuser@host-2171.customer.org'},{{{10,10,25,113},36701},'riakuser@host-2170.customer.org'}],{interval,#Ref<0.0.0.115>},false}

0xfffffd7fcd7ec590 Return addr 0x0000000004409830 (proc_lib:init_p_do_apply/3 + 56)
y(0)     riak_sysmon_filter
y(1)     {state,0,30,0,2,{0,nil},[{{{10,10,25,140},55002},'riakuser@host-2177.customer.org'},{{{10,10,25,90},53096},'riakuser@host-2167.customer.org'},{{{10,10,25,109},55747},'riakuser@host-2176.customer.org'},{{{10,10,25,88},39198},'riakuser@host-2173.customer.org'},{{{10,10,25,86},54204},'riakuser@host-2172.customer.org'},{{{10,10,25,112},52001},'riakuser@host-2169.customer.org'},{{{10,10,25,105},34719},'riakuser@host-2168.customer.org'},{{{10,10,25,117},44123},'riakuser@host-2171.customer.org'},{{{10,10,25,113},36701},'riakuser@host-2170.customer.org'}],{interval,#Ref<0.0.0.115>},false}
y(2)     riak_sysmon_filter
y(3)     <0.68.0>
y(4)     reset
y(5)     Catch 0x0000000005690648 (gen_server:handle_msg/5 + 1680)

0xfffffd7fcd7ec5c8 Return addr 0x0000000000676908 (<terminate process normally>)
y(0)     Catch 0x0000000004409850 (proc_lib:init_p_do_apply/3 + 88)

ok
(riakuser@host-2174.customer.org)13> 
```

**Comment by jonmeredith** [comment]
If affected by the issue, you will see the following in the logs

```
2012-03-19 17:41:49.780 [warning] <0.28966.1711> Mailbox size warning: 70094 messages: [{70094,<0.69.0>,riak_sysmon_filter}]
```

The process may be restarted by issuing this from the riak console (`riak attach` to log on `^D` to detach)

```
exit(whereis(riak_sysmon_filter), please_restart_now).
```

**Comment by rzezeski** [comment]
Is the issue that `net_kernel:get_status/3` calls `erlang:monitor_node/2`?  I'd like to understand the race condition.

**Comment by slfritchie** [comment]
IIRC it's a race between querying the ETS table and get_status() sending the message.  Monitoring the node isn't enough: if the recipient process is dead, you won't get an answer.  Or if there's a break in net distribution and the message is lost.

File: src/riak_sysmon_filter.erl [hunk 1]
```diff
@@ -338,24 +338,30 @@ annotate_dist_port(busy_dist_port, Port, S) ->
...
+             %% element(2, #net_address) is an {IpAddr, Port} tuple.
+             if element(1, T) == connection,
+                size(element(2, element(6, T))) == 2 ->
+                     {element(2, element(6, T)), element(2, T)};
+                true ->
+                     {bummer, bummer}
+                end
+         end || T <- ets:tab2list(sys_dist)]
+    catch _X:_Y ->
+            io:format("~p ~p @ ~p\n", [_X, _Y, erlang:get_stacktrace()]),
```
Comment (rzezeski) [inline src/riak_sysmon_filter.erl]: This should be a lager or error_logger call.

**Comment by rzezeski** [comment]
Verified that `riak_sysmon_filter:init/1` runs correctly and that the node map is correct.  Please change the `io:format` call and I'll +1 this.

```
(dev1@127.0.0.1)3> rp(sys:get_status(riak_sysmon_filter)).
{status,<0.71.0>,
        {module,gen_server},
        [[{'$ancestors',[riak_sysmon_sup,<0.69.0>]},
          {'$initial_call',{riak_sysmon_filter,init,1}}],
         running,<0.70.0>,[],
         [{header,"Status for generic server riak_sysmon_filter"},
          {data,[{"Status",running},
                 {"Parent",<0.70.0>},
                 {"Logged events",[]}]},
          {data,[{"State",
                  {state,0,30,0,2,
                         {0,nil},
                         [{{{127,0,0,1},52686},'dev4@127.0.0.1'},
                          {{{127,0,0,1},52659},'dev3@127.0.0.1'},
                          {{{127,0,0,1},52627},'dev2@127.0.0.1'}],
                         {interval,#Ref<0.0.0.110>},
                         false}}]}]]}
ok
```

**Pull request merged**